### PR TITLE
Duplicate some logic from LLVM's computeKnownBits

### PR DESCRIPTION
**Pull request by manasij7479**
Also improves pruning debug output

### CODE REVIEW

File: lib/Infer/DataflowPruning.cpp [hunk 1]
```diff
@@ -106,13 +133,41 @@ std::vector<ValueCache> DataflowPruningManager::generateInputSets(
...
+    if (I->K == souper::Inst::Var)
+      Cache[I] = {llvm::APInt(I->Width, 0xFFF)};
+  }
+  InputSets.push_back(Cache);
+
+  std::srand(0);
+  for (int i = 0 ; i < 5; ++i ) {
+    for (auto &&I : Inputs) {
+      if (I->K == souper::Inst::Var)
+        Cache[I] = {llvm::APInt(I->Width, std::rand() % llvm::APInt(I->Width, -1).getLimitedValue())};
```
Comment (regehr) [inline lib/Infer/DataflowPruning.cpp]: this does't work... you need a story for values larger than RAND_MAX

File: lib/Infer/DataflowPruning.cpp [hunk 2]
```diff
@@ -106,13 +133,41 @@ std::vector<ValueCache> DataflowPruningManager::generateInputSets(
...
   InputSets.push_back(Cache);
 
+  for (auto &&I : Inputs) {
+    if (I->K == souper::Inst::Var)
+      Cache[I] = {llvm::APInt(I->Width, 0xFFF)};
+  }
+  InputSets.push_back(Cache);
+
+  std::srand(0);
+  for (int i = 0 ; i < 5; ++i ) {
```
Comment (regehr) [inline lib/Infer/DataflowPruning.cpp]: I'd say avoid hard-coding stuff like this even in a really preliminary PR!

File: lib/Infer/DataflowPruning.cpp [hunk 3]
```diff
@@ -106,13 +133,41 @@ std::vector<ValueCache> DataflowPruningManager::generateInputSets(
...
+    for (auto &&I : Inputs) {
+      if (I->K == souper::Inst::Var)
+        Cache[I] = {llvm::APInt(I->Width, std::rand() % llvm::APInt(I->Width, -1).getLimitedValue())};
+    }
+    InputSets.push_back(Cache);
+  }
+
+  for (int i = 0 ; i < 5; ++i ) {
```
Comment (regehr) [inline lib/Infer/DataflowPruning.cpp]: why do you have the previous for loop and also this one?

File: lib/Infer/DataflowPruning.cpp [hunk 4]
```diff
@@ -106,13 +133,41 @@ std::vector<ValueCache> DataflowPruningManager::generateInputSets(
...
+    for (auto &&I : Inputs) {
+      if (I->K == souper::Inst::Var)
+        Cache[I] = {llvm::APInt(I->Width, std::rand() % 0xFFFFFFFF)};
+    }
+    InputSets.push_back(Cache);
+  }
```
Comment (regehr) [inline lib/Infer/DataflowPruning.cpp]: looks like you're not checking path conditions-- if you don't want to write that code, you need to leave a TODO or FIXME

File: lib/Infer/DataflowPruning.cpp [hunk 5]
```diff
@@ -106,13 +133,41 @@ std::vector<ValueCache> DataflowPruningManager::generateInputSets(
...
+      if (I->K == souper::Inst::Var)
+        Cache[I] = {llvm::APInt(I->Width, std::rand() % llvm::APInt(I->Width, -1).getLimitedValue())};
+    }
+    InputSets.push_back(Cache);
+  }
+
+  for (int i = 0 ; i < 5; ++i ) {
+    for (auto &&I : Inputs) {
+      if (I->K == souper::Inst::Var)
+        Cache[I] = {llvm::APInt(I->Width, std::rand() % 0xFFFFFFFF)};
```
Comment (regehr) [inline lib/Infer/DataflowPruning.cpp]: what's the purpose of the modulus operator?

File: lib/Infer/Interpreter.cpp [hunk 6]
```diff
@@ -2,29 +2,27 @@
 
 namespace souper {
 
-  bool hasReservedHelper(Inst *I, std::set<Inst *> &Visited) {
-    if (I->K == Inst::ReservedConst || I->K == Inst::ReservedInst)
+  bool hasReservedHelper(Inst *I, std::set<Inst *> &Visited, bool C, bool H) {
```
Comment (regehr) [inline lib/Infer/Interpreter.cpp]: please use better names than C and H

File: lib/Infer/DataflowPruning.cpp (hunk 3 above)
Comment (manasij7479) [reply lib/Infer/DataflowPruning.cpp]: This one chooses small inputs on purpose.

File: lib/Infer/DataflowPruning.cpp (hunk 5 above)
Comment (manasij7479) [reply lib/Infer/DataflowPruning.cpp]: Choose small inputs, to prune shifts better.
Changed 0xFFFFFFFF to I->Width

File: lib/Infer/DataflowPruning.cpp (hunk 1 above)
Comment (manasij7479) [reply lib/Infer/DataflowPruning.cpp]: What do you suggest?
(Probably fine to keep this for now, and switch to C++ <random> for experiments)

**Comment by regehr** [comment]
thanks!

**Pull request merged**